base/scopedvalues.jl

```diff
@@ -2,9 +2,57 @@
 
 module ScopedValues
 
-export ScopedValue, with, @with
+export ScopedValue, LazyScopedValue, with, @with
 public get
 
+"""
+    AbstractScopedValue{T}
+
+Abstract base type for scoped values that propagate values across
+dynamic scopes. All scoped value types must extend this abstract type.
+
+See also: [`ScopedValue`](@ref), [`LazyScopedValue`](@ref)
+
+!!! compat "Julia 1.13"
+    AbstractScopedValue requires Julia 1.13+.
+"""
+abstract type AbstractScopedValue{T} end
+
+
+"""
+    LazyScopedValue{T}(f::OncePerProcess{T})
+
+A scoped value that uses an `OncePerProcess{T}` to lazily compute its default value
+when none has been set in the current scope. Unlike `ScopedValue`, the default is
+not evaluated at construction time but only when first accessed.
+
+# Examples
+
+```julia-repl
+julia> using Base.ScopedValues;
+
+julia> const editor = LazyScopedValue(OncePerProcess(() -> ENV["JULIA_EDITOR"]));
+
+julia> editor[]
+"vim"
+
+julia> with(editor => "emacs") do
+           sval[]
+       end
+"emacs"
+
+julia> editor[]
+"vim"
+```
+
+!!! compat "Julia 1.13"
+    LazyScopedValue requires Julia 1.13+.
+"""
+mutable struct LazyScopedValue{T} <: AbstractScopedValue{T}
+    const getdefault::OncePerProcess{T}
```

Comment:
It it worth abstracting over what OncePer is used here (just use `F <: Function`)? We could potentially consider also using `Returns{T}(default)` to re-combine LazyScopedValue/ScopedValue, though I don't know if that that is important. There is not (and cannot be) a convert method for OncePerProcess{S} to OncePerProcess{T}, so it could be a bit tedious sometimes if users are trying to write the type parameters explicitly while also using OncePerProcess's auto-interferred parameter ability (though users really shouldn't be writing the LazyScopedValue apply_type explicitly)

Reply:
Sure, makes sense. I think flexibility would make more sense with alternatives like OncePerDepot than the ones Valentin mentioned or OncePerPeriod (aka debounce or throttle). But as you say, this is conservatively what is actually needed now